x-pack/legacy/plugins/task_manager/lib/identify_es_error.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+interface ESErrorCausedBy {
+  reason?: string;
+  caused_by?: ESErrorCausedBy;
+}
+
+interface ESError {
+  root_cause?: ESErrorCausedBy[];
+  caused_by?: ESErrorCausedBy;
+}
+
+function extractCausedByChain(
+  causedBy: ESErrorCausedBy = {},
+  accumulator: string[] = []
+): string[] {
+  const { reason, caused_by: innerCausedBy } = causedBy;
+
+  if (reason) {
+    accumulator.push(reason);
+  }
+
+  if (innerCausedBy) {
+    return extractCausedByChain(innerCausedBy, accumulator);
+  }
+
+  return accumulator;
+}
+
+/**
+ * Identified causes for ES Error
+ *
+ * @param err Object Error thrown by ES JS client
+ * @return ES error cause
+ */
+export function identifyEsError(err: { response: string }) {
+  const { response } = err;
+
+  if (response) {
+    const { error } = JSON.parse(response) as { error?: ESError };
+    if (error) {
+      const { root_cause: rootCause = [], caused_by: causedBy } = error;
+
+      // The caused_by chain has the most information so use that if it's available. If not then
+      // settle for the root_cause.
+      const causedByChain = extractCausedByChain(causedBy);
+      if (causedByChain.length) {
```

Comment:
I would have expected it to traverse the complete `caused_by` and `root_clause` tree of messages, but I'm unfamiliar with these ES error trees.

Reply:
I thought it was?
This code is actually modified from another place in Kibana where we do something similar, but my understanding is that it does go deep as `extractCausedByChain` goes recursively.

Reply:
Cleaned this up to make it clearer